src/Protocol/ActivityPub/Receiver.php

```diff
@@ -2025,4 +2027,137 @@ private static function hasArrived(string $id): bool
+			Logger::notice('Empty type', ['activity' => $ldactivity, 'uid' => $uid]);
+			return;
+		}
+
+		$object_id   = JsonLD::fetchElement($ldactivity, 'as:object', '@id') ?? '';
+		$object_type = self::fetchObjectType($ldactivity, $object_id, $uid);
+		if (!$object_type && !$object_id) {
+			Logger::notice('Empty object type or id', ['activity' => $ldactivity, 'uid' => $uid]);
+			return;
+		}
+
+		Logger::debug('Processing activity', ['type' => $type, 'object_type' => $object_type, 'object_id' => $object_id, 'activity' => $ldactivity]);
+		self::routeC2SActivities($type, $object_type, $object_id, $uid, $application, $ldactivity);
+		throw new \Friendica\Network\HTTPException\AcceptedException();
+	}
+
+	private static function getTargets(array $object, string $actor): array
+	{
+		$profile   = APContact::getByURL($actor);
+		$followers = $profile['followers'];
+
+		$targets = [];
+
+		foreach (['as:to', 'as:cc', 'as:bto', 'as:bcc'] as $element) {
+			switch ($element) {
+				case 'as:to':
+					$type = self::TARGET_TO;
+					break;
+				case 'as:cc':
+					$type = self::TARGET_CC;
+					break;
+				case 'as:bto':
+					$type = self::TARGET_BTO;
+					break;
+				case 'as:bcc':
+					$type = self::TARGET_BCC;
+					break;
+			}
+			$receiver_list = JsonLD::fetchElementArray($object, $element, '@id');
+			if (empty($receiver_list)) {
+				continue;
+			}
+
+			foreach ($receiver_list as $receiver) {
+				if ($receiver == self::PUBLIC_COLLECTION) {
+					$targets[self::TARGET_GLOBAL] = ($element == 'as:to');
+					continue;
+				}
+
+				if ($receiver == $followers) {
+					$targets[self::TARGET_FOLLOWER] = true;
+					continue;
+				}
+				$targets[$type][] = Contact::getIdForURL($receiver);
+			}
+		}
+		return $targets;
+	}
+
+	private static function routeC2SActivities(string $type, string $object_type, string $object_id, int $uid, array $application, array $ldactivity)
+	{
+		switch ($type) {
+			case 'as:Create':
+				if (in_array($object_type, self::CONTENT_TYPES)) {
+					self::createContent($uid, $application, $ldactivity);
+				}
+				break;
+			case 'as:Update':
+				if (in_array($object_type, self::CONTENT_TYPES) && !empty($object_id)) {
+					self::updateContent($uid, $object_id, $application, $ldactivity);
+				}
+				break;
+			case 'as:Follow':
+				if (in_array($object_type, self::ACCOUNT_TYPES) && !empty($object_id)) {
+					self::followAccount($uid, $object_id, $ldactivity);
+				}
+				break;
+		}
+	}
+
+	private static function createContent(int $uid, array $application, array $ldactivity)
+	{
+		$object_data = self::processObject($ldactivity['as:object'], true);
+		$item = Processor::processC2SContent($object_data, $application, $uid);
+		Logger::debug('Got data', ['item' => $item, 'object' => $object_data]);
+
+		$id = Item::insert($item, true);
+		if (!empty($id)) {
+			$item = Post::selectFirst(['uri-id'], ['id' => $id]);
+			if (!empty($item['uri-id'])) {
+				System::jsonExit(Transmitter::createActivityFromItem($id));
```

Comment:
As said at the throw. I think the only place, where we should exit a HTTP Request should be inside a `BaseModule` based class. The "inner" classes should return a value which can guide the caller what to do.

Another thing is that with returning something instead of calling exit or throwing exceptions, we can specifically can test classes :)

Reply:
I agree with that, the `exit` methods should only be called in Module classes.